### PR TITLE
compression level change with config option

### PR DESCRIPTION
**Pull request by Leguan16**
Supersedes #9703 

As from the author requested: Here a new version with the correct patches and a config option.

The config option is located under the `misc` section. I did not know if I should change the version of the config but that should not stop this PR from being merged. 

Is it possible to add people as co-authors so the original author also gets credits?

### CODE REVIEW

File: patches/server/0005-Paper-config-files.patch [hunk 1]
```diff
@@ -465,7 +465,7 @@ index 0000000000000000000000000000000000000000..45f1436cdd4b81b621ab71e4336c2aa6
 +
 +@SuppressWarnings({"CanBeFinal", "FieldCanBeLocal", "FieldMayBeFinal", "NotNullFieldNotInitialized", "InnerClassMayBeStatic"})
 +public class GlobalConfiguration extends ConfigurationPart {
-+    static final int CURRENT_VERSION = 28;
++    static final int CURRENT_VERSION = 29;
```
Comment (Owen1212055) [inline patches/server/0005-Paper-config-files.patch]: Don't think we need to update the version here

File: patches/server/0005-Paper-config-files.patch [hunk 2]
```diff
@@ -725,6 +725,7 @@ index 0000000000000000000000000000000000000000..45f1436cdd4b81b621ab71e4336c2aa6
 +        public boolean useAlternativeLuckFormula = false;
 +        public boolean useDimensionTypeForCustomSpawners = false;
 +        public boolean strictAdvancementDimensionCheck = false;
++        public int compressionLevel = -1;
```
Comment (lynxplay) [inline patches/server/0005-Paper-config-files.patch]: Would probably be better to use IntOrDefault here.
-1 yields default compression for the zip deflater as well as for the native ones velocity implements (6 iirc)

**Review by lynxplay — CHANGES_REQUESTED** [review]

File: patches/server/0005-Paper-config-files.patch (hunk 1 above)
Comment (lynxplay) [reply patches/server/0005-Paper-config-files.patch]: We don't yea. Version bumps are  only needed if we need a migration to be run on existing data.

**Comment by lynxplay** [comment]
You can generally add a coauthor by simply adding their name into the patch, see e.g. https://github.com/PaperMC/Paper/blob/master/patches/server/0872-Fix-a-bunch-of-vanilla-bugs.patch#L60

**Comment by Leguan16** [comment]
the placement of the option is ok or should I move that as well?

File: patches/server/0005-Paper-config-files.patch (hunk 1 above)
Comment (Leguan16) [reply patches/server/0005-Paper-config-files.patch]: alright thanks!

**Comment by lynxplay** [comment]
> the placement of the option is ok or should I move that as well?

yea :+1: imo it is.

File: patches/server/0005-Paper-config-files.patch (hunk 2 above)
Comment (Leguan16) [reply patches/server/0005-Paper-config-files.patch]: Ok so I have tried this and I don't seem to be able to do that.
It tells me that `No TypeSerializer found for field compressionLevel of type class io.papermc.paper.configuration.type.IntOr$Default`, before it just did not want to create the `misc` section at all.
To my eyes I have made everything like where it is already in use. 
The updated `GlobalConfiguration` variable:
```java
public IntOr.Default compressionLevel = IntOr.Default.USE_DEFAULT
```
The updated line in `Connection#setupCompression()`:
```java
[...].create(io.papermc.paper.configuration.GlobalConfiguration.get().misc.compressionLevel.or(-1));
```
Do I have to change anything else besides that?

Comment (Leguan16) [reply patches/server/0005-Paper-config-files.patch]: Ah so i think this is why:
https://github.com/PaperMC/Paper/blob/master/patches/server/0005-Paper-config-files.patch#L1166
(only the world-config has the IntOr Serializers registered)
What are we doing now? Should we create a separate PR adding the stuff to the global config or should I include that here?

Comment (lynxplay) [reply patches/server/0005-Paper-config-files.patch]: include the registration here :+1:

Comment (Leguan16) [reply patches/server/0005-Paper-config-files.patch]: should be fixed now.

**Review by lynxplay — APPROVED** [review]

**Review by kennytv — APPROVED** [review]

**Comment by PedroMPagani** [comment]
@Leguan16 good job ;), it'd be nice if paper added on the docs examples of the available compression options with example of rates, as a benchmark of course, this should benefit a few servers a lot with bandwidth decreasing as well as more performance and cost decreasing :)

**Comment by lynxplay** [comment]
This does require a PR to the docs before it may be merged yea. @Leguan16 if you have the time, it would be great if you can propose the change there as well :+1:

**Comment by Leguan16** [comment]
I can add a description to the `global-configuration.mdx` what it does but I have no time to benchmark this and create a proper docs page for it. Besides, I don't even know how I should benchmark this. 
@PedroMPagani you seem to know what it does exactly so what should I put in the description (So I have it correct)?

**Comment by lynxplay** [comment]
I don't see the usage for benchmarking on the docs. A quick explanation that a higher compression level means more CPU time spend for less data transmitted and a lower value meaning less CPU time spend but more data transmitted is fine.

**Comment by Leguan16** [comment]
Alright! I will add that later.

**Comment by Leguan16** [comment]
I opened a PR in the docs repo
https://github.com/PaperMC/docs/pull/242

**Comment by olijeffers0n** [comment]
Docs is ready to go 😄

**Pull request merged**